crates/ty_python_semantic/src/types/function.rs

```diff
@@ -0,0 +1,857 @@
+use std::str::FromStr;
+
+use bitflags::bitflags;
+use ruff_db::diagnostic::Span;
+use ruff_db::files::{File, FileRange};
+use ruff_python_ast as ast;
+use ruff_text_size::Ranged;
+
+use crate::module_resolver::{KnownModule, file_to_module};
+use crate::semantic_index::ast_ids::HasScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::semantic_index;
+use crate::semantic_index::symbol::ScopeId;
+use crate::symbol::{Boundness, Symbol, symbol_from_bindings};
+use crate::types::generics::GenericContext;
+use crate::types::narrow::ClassInfoConstraintFunction;
+use crate::types::signatures::{CallableSignature, Signature};
+use crate::types::{BoundMethodType, CallableType, Type, TypeMapping, TypeVarInstance};
+use crate::{Db, FxOrderSet};
+
+/// A collection of useful spans for annotating functions.
+///
+/// This can be retrieved via `FunctionType::spans` or
+/// `Type::function_spans`.
+pub(crate) struct FunctionSpans {
+    /// The span of the entire function "signature." This includes
+    /// the name, parameter list and return type (if present).
+    pub(crate) signature: Span,
+    /// The span of the function name. i.e., `foo` in `def foo(): ...`.
+    pub(crate) name: Span,
+    /// The span of the parameter list, including the opening and
+    /// closing parentheses.
+    #[expect(dead_code)]
+    pub(crate) parameters: Span,
+    /// The span of the annotated return type, if present.
+    pub(crate) return_type: Option<Span>,
+}
+
+bitflags! {
+    #[derive(Copy, Clone, Debug, Eq, PartialEq, Default, Hash)]
+    pub struct FunctionDecorators: u8 {
+        /// `@classmethod`
+        const CLASSMETHOD = 1 << 0;
+        /// `@typing.no_type_check`
+        const NO_TYPE_CHECK = 1 << 1;
+        /// `@typing.overload`
+        const OVERLOAD = 1 << 2;
+        /// `@abc.abstractmethod`
+        const ABSTRACT_METHOD = 1 << 3;
+        /// `@typing.final`
+        const FINAL = 1 << 4;
+        /// `@typing.override`
+        const OVERRIDE = 1 << 6;
+    }
+}
+
+bitflags! {
+    /// Used for the return type of `dataclass_transform(…)` calls. Keeps track of the
+    /// arguments that were passed in. For the precise meaning of the fields, see [1].
+    ///
+    /// [1]: https://docs.python.org/3/library/typing.html#typing.dataclass_transform
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, salsa::Update)]
+    pub struct DataclassTransformerParams: u8 {
+        const EQ_DEFAULT = 0b0000_0001;
```

Comment:
Nit: Should we use `1 << 0` similar to `FunctionDecorators` to have a single style for defining bitflags